### ISSUE TITLE
Mod tap for home row modifiers.

### ISSUE THREAD

**Issue by Maverickman13** [WootingKb/wootility-issues#309]
**Is your feature request related to a problem? Please describe.**
I want to use home-row modifiers, but for general typing, the delay is disregarded if a modifier and another key is pressed at same time. 

**Describe the solution you'd like**
Allow a toggle to not have modifier keys automatically enable when another key is pressed at the same time. 

**Platform information**
Win 11. 

**Wootility and Firmware version**
Wootility Web v 5.0.0 beta-2

**Keyboard Type**
Wooting 80HE

**Comment by diogotr7** [comment]
The latest Wootility v5.1-beta.0 and accompanying firmware added support for this. Please test and give feedback!

**Comment by YourSandwich** [comment]
Any recommended settings? When I set mod tap, taped keystrokes have too much of a latency even if I optimize the delay and tick, I seem to not get it correctly.

**Comment by MauriceMatthys** [comment]
> The latest Wootility v5.1-beta.0 and accompanying firmware added support for this. Please test and give feedback!

@diogotr7 
How do we best give feedback? Because I saw the announcement that support for HRM was added back when it released, but I quickly turned it back off since the ModTap key still waits if a key is pressed while the MT is active and then only triggers after the key that was pressed after is released.

For example Online is turned into Onilne if you're using L as a MT key. That's also the behavior you see when using the Analog monitor.

**Comment by bapeey** [comment]
In my opinion, if the hold time hasn't passed when another key is pressed, the key with the modifier should be sent first before the other key. With the current implementation, the key with the modifier is sent after the second key, and you have to type slowly to prevent that from happening

**Comment by simon-wh** [comment]
@MauriceMatthys @bapeey Just to double check, are you both changing this setting to enable the Home Row mod use case? 

<img width="803" alt="Image" src="https://github.com/user-attachments/assets/1a66616d-f899-4166-af7c-06fca48aee98" />

**Comment by bapeey** [comment]
Yep, I have the 's' with ModTap, and as you can ese in this mesasge the 's' come sout after the next presed key

-----
I' tested Kanata with this [config](https://[raw.githubusercontent.com/dreamsofcode-io/home-row-mods/refs/heads/main/kanata/linux/kanata.kbd](https://raw.githubusercontent.com/dreamsofcode-io/home-row-mods/refs/heads/main/kanata/linux/kanata.kbd)) and not have the issue, is like the keys are getting buffered

**Comment by MauriceMatthys** [comment]
@simon-wh 

![Image](https://github.com/user-attachments/assets/b5afce97-322e-48e1-be8a-4461db578fb6)l

You can see it pretty clearly using the analog monitor. Despite typing online it comes out as onilne because even though the l key is pressed (1) before the i key, it only activates (2) after and deactivates once the key is fully lifted (3).

**Comment by farnoy** [comment]
I'm new to wooting, what are these analog monitors you use for debugging? I flashed my 80HE to `2.1.0-beta.2` and the new JIT setting is still insufficient. With it set to "Never", I need to press my modifiers first (like ctrl+shift), then my mod tap key (I use the L-Super key position, activating fn layer 2 when held) and then it has the combined effect of ctrl+shift+arrow keys. With JIT set to Always, it removes the delay after pressing ctrl+shift+ModTap and before being able to press the hjkl arrow keys. That means I can press ctrl+shift+modtap+h as fast as I'm able to, but it only works in that order of presses.

Nothing I tried lets me press the ModTap key first, then add modifiers and finally use hjkl to send arrow keys. This is most useful when I navigate words with ctrl+modtap+(h/j/k/l) but then add shift into the mix to start selecting words. On QMK, I just used a MO(layer) and it worked perfectly, can't seem to emulate this as well with ModTap on Wooting?

**Comment by farnoy** [comment]
Figured it out right after posting... Sorry about the noise. My issue was the ctrl & shift keys on the upper layer were blank, which I thought meant transparent like in QMK. I think it works very well for my usecase now with JIT set to "Only modifiers"

**Comment by Scott-Meyer** [comment]
Is there any update on this? The inability to use home row mods prevents me from having this as my only keyboard.

**Comment by attole** [comment]
Are there any updates on this? 

I’m running into a the same issue, described by @MauriceMatthys. Home Row Mod (HRM) Taps (`JIT = Never`) still breaks fast typing. Current version: `Wootility Web v5.2.1`, `Firmware: v2.12.2`.

**Issue summary**

With `JIT = Never`, a Mod Tap (MT) key does not emit the tap until the key is released. If another key is pressed while the MT is still physically held, even for a very short overlap, the tap is delayed and emitted after key is released, not before next input.

This causes reordered characters during rolling/burst typing (very fast typing sequences, using different fingers for different keys at the same time, where the next key is pressed before the previous one is fully released, having overlapping key presses).

What actually happens, given the previous example (`Online` is turned into `Onilne` if you're using `L` as a MT key):

1. Press `L` (MT key)
2. Quickly press `I` (default key)
3. `I` is sent immediately
4. Both keys released
5. `L` is sent after release 
6. Result is `il`, not `li`

All of this occurs well below the hold threshold. The behavior is not threshold-dependent; the tap is gated on full key release. Until the MT key is released, no tap event is emitted, regardless of how short the press is. From the analog monitor, sequence is:

- MT key is pressed first
- It only activates after the next key
- Deactivates only on full release

Mod Keys on HRM rely on overlapping presses and rolling motion. If the Mod Key is pressed long enough (beyond threshold) it is Mod key, but in any other cases it should be threated as default tap (and be emitted before any other input, be sequential).

This makes HRM implementations on current MT keys (with `JIT = Never`) unusable for fast typists, while `JIT = Always / Only Modifiers` causes premature modifier activation.

**Suggested improvement**

Change the current implementation of `JIT = Never` behavior to immediately emit the tap instead of waiting for release, when another key is pressed while a MT is held and the hold threshold has not been reached.

This preserves HRM intent while matching fast typing behavior and preventing reordered characters.

Is there any progress or planned change around this?
Tagging for visibility: @simon-wh @diogotr7

**Comment by simon-wh** [comment]
[The latest beta update `v5.3.0-beta.2`](https://wooting.io/wootility/changelogs/5.3.0-beta.2) includes a new option to the Mod Tap JIT setting called "Tap Key". This will input the Tap key immediately when another key is pressed while the activate Mod Tap key state is still indeterminate. It will also ensure that the two keys don't first appear up on the same input report, ensuring the text input ordering is correct.

In my testing this allowed me to type normally with Home row mods configured. Let me know how if this is working well for you now 🙏

**Comment by YourSandwich** [comment]
@simon-wh Thank you, this now works as intended. Need to get used to the latency, though. Set it to 200ms, 120ms felt much better, but the keys were not typing if I was too slow. I guess this is something I need to balance out.

**Comment by MauriceMatthys** [comment]
@simon-wh There's no longer any errors when using the new settings and it seems to be in a working state. Thanks 😄.

I did notice the typing is a bit "choppy?" when going faster than the delay. Sticking to my example of online with every home key  being a mod-tap, on the left is me typing slow and on the right I'm typing fast.

<img width="1920" height="993" alt="Image" src="https://github.com/user-attachments/assets/b96f8841-e083-4562-8b49-73c2cb49fc9e" />

You can see on the right the L activation only happens just after pressing the i key.

I don't remember if this happened with my kanata config (it probably did because i can't imagine how it would work otherwise) to get homerow mods in software and I'm not an expert on keyboard software so I think this might just be inherent to the JIT behavior since it has to wait to check if another key is pressed or not. 
So I'll probably have to get used to it.

As a last thing this is just a feature request but Kanata allows you to still hold the tap key itself using double-tap + hold. So if someone was using homerow mods and wanted to spam "fffffffffffff" like so they just held the f key down, let it go and then keep holding it down as long as they wanted to keep typing f for. If that could be implemented as a toggle that would be nice QoL.

But the Tap Key JIT behavior solves 99% of my issues so I'm happy already.

**Comment by artyultra** [comment]
@MauriceMatthys i see this post is only month old but im still having this problem and i dont see a the tap setting in jit

**Comment by artyultra** [comment]
@MauriceMatthys oh disregard i saw the part about beta i just installed it

**Comment by attole** [comment]
@simon-wh Thanks, that fixed it!

**Comment by pluskid** [comment]
> As a last thing this is just a feature request but Kanata allows you to still hold the tap key itself using double-tap + hold. So if someone was using homerow mods and wanted to spam "fffffffffffff" like so they just held the f key down, let it go and then keep holding it down as long as they wanted to keep typing f for. If that could be implemented as a toggle that would be nice QoL.

+1 to this feature, this is also standard feature in QMK / ZMK. This is especially useful for vim and other editor that uses hjkl for cursor movement. Not being able to hold down the key to move around makes it very hard to use them.
